model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserLoginFailureProviderFactory.java

```diff
@@ -0,0 +1,175 @@
+import org.keycloak.models.sessions.infinispan.remotestore.RemoteCacheSessionListener;
+import org.keycloak.models.sessions.infinispan.remotestore.RemoteCacheSessionsLoader;
+import org.keycloak.models.sessions.infinispan.util.InfinispanUtil;
+import org.keycloak.models.sessions.infinispan.util.SessionTimeouts;
+import org.keycloak.models.utils.KeycloakModelUtils;
+import org.keycloak.models.utils.PostMigrationEvent;
+
+import java.io.Serializable;
+import java.util.Set;
+import java.util.function.BiFunction;
+
+public class InfinispanUserLoginFailureProviderFactory implements UserLoginFailureProviderFactory {
+
+    private static final Logger log = Logger.getLogger(InfinispanUserLoginFailureProviderFactory.class);
+
+    public static final String PROVIDER_ID = "infinispan";
+
+    public static final String REMOVE_ALL_LOGIN_FAILURES_EVENT = "REMOVE_ALL_LOGIN_FAILURES_EVENT";
+
+    private Config.Scope config;
+
+    private RemoteCacheInvoker remoteCacheInvoker;
+
+    @Override
+    public UserLoginFailureProvider create(KeycloakSession session) {
+        InfinispanConnectionProvider connections = session.getProvider(InfinispanConnectionProvider.class);
+        Cache<LoginFailureKey, SessionEntityWrapper<LoginFailureEntity>> loginFailures = connections.getCache(InfinispanConnectionProvider.LOGIN_FAILURE_CACHE_NAME);
+
+        return new InfinispanUserLoginFailureProvider(session, remoteCacheInvoker, loginFailures);
+    }
+
+    @Override
+    public void init(Config.Scope config) {
+        this.config = config;
+    }
+
+    @Override
+    public void postInit(final KeycloakSessionFactory factory) {
+        KeycloakModelUtils.runJobInTransaction(factory, (KeycloakSession session) -> {
+            this.remoteCacheInvoker = new RemoteCacheInvoker();
+        });
+
+        factory.register(event -> {
+            if (event instanceof PostMigrationEvent) {
+                KeycloakModelUtils.runJobInTransaction(factory, (KeycloakSession session) -> {
+                    checkRemoteCaches(session);
+                    registerClusterListeners(session);
+                    loadSessionsFromRemoteCaches(session);
+                });
+            }
+        });
+    }
+
+    protected void registerClusterListeners(KeycloakSession session) {
+        KeycloakSessionFactory sessionFactory = session.getKeycloakSessionFactory();
+        ClusterProvider cluster = session.getProvider(ClusterProvider.class);
+
+        cluster.registerListener(REMOVE_ALL_LOGIN_FAILURES_EVENT,
+                new AbstractUserSessionClusterListener<RemoveAllUserLoginFailuresEvent, UserLoginFailureProvider>(sessionFactory, UserLoginFailureProvider.class) {
+
+            @Override
+            protected void eventReceived(KeycloakSession session, UserLoginFailureProvider provider, RemoveAllUserLoginFailuresEvent sessionEvent) {
+                provider.removeAllUserLoginFailures(sessionEvent.getRealmId());
```

Comment:
This has to remove only *local* login failures

Reply:
Explained previously and it's now calling directly `removeAllLocalUserLoginFailuresEvent`.